### DISCUSSION TITLE
Question about HybridCache value type

### DISCUSSION THREAD

**Discussion by Yiling-J** [facebook/CacheLib#232] [Q&A] [answered]
The HybridCache part of CacheLib is very intersting and I'm considering implementing similar thing in my Go cache package. I'm not expert of C++ and can't find answer in documentation, so I think ask here is the best. So my question is, does HybridCache support any type of value? From the documentation I think HybridCache will copy memory data to NVM, what if value is a struct contains pointer or vector members, will HybridCache copy pointer address? Currently my Go package is generic so value type can be anything, I think I need to limit value types to make HybridCache work in Go.

**Comment by haowu14** [comment, marked as answer]
Hybrid cache (and traditional cachelib) doesn't support types. It's just byte strings so no pointers allowed there.

Object-cache is only used for DRAM cache (no flash) and you can cache c++ objects.

**Reply by Yiling-J** [reply]
Thanks for the explanation! Seems like I can't find documentations about Object-cache, is that objcache2 in experimental? And Object-cache is only used for DRAM cache because it's not safe to marshal object to byte? BTW cachelib's trace and Architecture Guide are very useful! I already use the trace in my cache package benchmark and try to implement hybrid cache based on the Architecture Guide.

**Reply by haowu14** [reply]
Before Object-Cache was introduced, the users would serialize their customized datastructure into byte strings and use CacheLib. For some users, this serialization and deserialization is too expensive, especially for users with frequent updates. Therefore Object-Cache was introduced, aiming to reduce this cost.

But if we are going to move data to flash, we'd need to serialized the data anyway, so we might as well just use regular CacheLib.

**Reply by Yiling-J** [reply]
@haowu14 Actually I think Object-Cache with hybrid has one benefit, data are deserialized and insert to memory before return to client, which means hit memory part of hybrid cache will avoid the cost of deserialization. If deserialization is expensive this can improve performance compared to regular byte cache.

**Comment by pbhandar2** [comment]
Closing this discussion as it has been inactive. If this is still relevant, then please create a new discussion post and we will get back to you as soon as we can. Thank you!
